saliency/tf1/utils.py

```diff
@@ -0,0 +1,142 @@
+
+
+def create_tf1_call_model_function(graph,
+                                   session,
+                                   y=None,
+                                   x=None,
+                                   conv_layer=None):
+  """Creates a call_model_function which calls the TF1 model specified.
+
+  Args:
+    graph: The TensorFlow graph to evaluate masks on.
+    session: The current TensorFlow session.
+    y: (Optional) The output tensor of the model. This tensor should be of size
+      1.
+    x: The input tensor of the model. The outer dimension should be the batch
+      size.
+    conv_layer: (Optional) The convolution layer tensor of the model. The outer
+      dimension should be the batch size.
+
+  Raises:
+    ValueError: If input tensor x is missing.
+
+  Returns:
+    call_model_function: A function with the following signature:
+        call_model_function(x_value_batch,
+                              call_model_args=None,
+                              expected_keys=None):
+          x_value_batch - Input for the model, given as a batch (i.e. dimension
+            0 is the batch dimension, dimensions 1 through n represent a single
+            input).
+          call_model_args - Other arguments used to call and run the model.
+          expected_keys - List of keys that are expected in the output.
+  """
+  _import_tf()
+
+  with graph.as_default():
+    if x is None:
+      raise ValueError('Expected input tensor for x but is equal to None.')
+    if y is not None:
+      output_gradients = tf.compat.v1.gradients(y, x)[0]
+      if conv_layer is not None:
+        conv_gradients = tf.compat.v1.gradients(y, conv_layer)[0]
+
+  def convert_keys_to_fetches(expected_keys):
+    """Converts expected keys into an array of fetchable tensors.
+
+    Args:
+      expected_keys: Array of strings, representing the expected keys.
+
+    Raises:
+        RuntimeError: If tensor required for expected_key doesn't exist.
+
+    Returns:
+      Array of fetches that can be used in a session.run call.
+    """
+    fetches = []
+    y_tensor_required = [OUTPUT_LAYER_VALUES, INPUT_OUTPUT_GRADIENTS,
+                         CONVOLUTION_OUTPUT_GRADIENTS]
+    conv_tensor_required = [CONVOLUTION_LAYER_VALUES, 
+                            CONVOLUTION_OUTPUT_GRADIENTS]
+    for expected_key in expected_keys:
+      if expected_key in y_tensor_required:
+        if y is None:
+          raise RuntimeError(MISSING_Y_ERROR_MESSAGE.format(expected_key))
+        if expected_key == OUTPUT_LAYER_VALUES:
+          fetches.append(y)
+        elif expected_key == INPUT_OUTPUT_GRADIENTS:
+          fetches.append(output_gradients)
+      if expected_key in conv_tensor_required:
+        if conv_layer is None:
+          raise RuntimeError(MISSING_CONV_LAYER_ERROR_MESSAGE.format(
+              expected_key))
+        if expected_key == CONVOLUTION_LAYER_VALUES:
+          fetches.append(conv_layer)
+        else:
```

Comment:
You'll have to explain what you want here. How should we handle a potentially missing conv_layer? Are there nested ifs so we can validate each key? i.e.

```
if key=y:
   fetches.append()
...
elif key=convolution_layer_values:
   if conv_layer is not None:
       fetches.append()
   else:
        raise RuntimeError
else:
   raise ValueError('invalid expected key')
```

Reply:
Yes, let's just fail if the requested key cannot be added. I just want to avoid the situation that someone adds a new key to the VALID_EXPECTED_KEYS list but forgets to update the convert_keys_to_fetches
 method. As a result, the key will be silently ignore and the code that calls the method may crash.

Reply:
Right, that makes sense. Changed to put each key in a separate if clause.

Reply:
Ah I see, yes it's good to avoid that check requiring multiple changes at once in different places. Changed.